Partial fix for hangs.

commit 21ccf5ed07ed7b9f208827dc7ba9b932bfa6cfaa
Author: Dmitry V. Levin <ldv@altlinux.org>
Date:   Thu Jan 13 14:49:39 2011 +0000

    Test PTRACE_O_TRACECLONE and PTRACE_O_TRACEVFORK along with PTRACE_O_TRACEFORK
    
    * strace.c (test_ptrace_setoptions): Add PTRACE_O_TRACECLONE and
    PTRACE_O_TRACEVFORK to PTRACE_SETOPTIONS call, to test exactly
    the same set of options that is going to be used later in trace().

diff --git a/strace.c b/strace.c
index e1ae8f1..6b0ebac 100644
--- a/strace.c
+++ b/strace.c
@@ -726,13 +726,14 @@ test_ptrace_setoptions(void)
 				kill(tracee_pid, SIGKILL);
 		}
 		else if (WIFSTOPPED(status)) {
+			const unsigned int test_options = PTRACE_O_TRACECLONE |
+							  PTRACE_O_TRACEFORK |
+							  PTRACE_O_TRACEVFORK;
 			if (status >> 16 == PTRACE_EVENT_FORK)
-				ptrace_setoptions |= (PTRACE_O_TRACEVFORK |
-						      PTRACE_O_TRACECLONE |
-						      PTRACE_O_TRACEFORK);
+				ptrace_setoptions |= test_options;
 			if (WSTOPSIG(status) == SIGSTOP) {
 				if (ptrace(PTRACE_SETOPTIONS, pid, NULL,
-					   PTRACE_O_TRACEFORK) < 0) {
+					   test_options) < 0) {
 					kill(pid, SIGKILL);
 					return -1;
 				}
